onboard/catkin_ws/src/controls/src/controls_utils.cpp

```diff
@@ -0,0 +1,465 @@
+
+        // Read desired power limits
+        YAML::Node desired_power_limits_node = config["desired_power_limits"];
+        for (const auto &axis : AXES) {
+            desired_power_min[axis] = desired_power_limits_node[AXES_NAMES.at(axis)]["min"].as<double>();
+            desired_power_max[axis] = desired_power_limits_node[AXES_NAMES.at(axis)]["max"].as<double>();
+        }
+
+        // Read static power vector
+        YAML::Node static_power_global_node = config["static_power_global"];
+        static_power_global.setX(static_power_global_node[AXES_NAMES.at(AxesEnum::X)].as<double>());
+        static_power_global.setY(static_power_global_node[AXES_NAMES.at(AxesEnum::Y)].as<double>());
+        static_power_global.setZ(static_power_global_node[AXES_NAMES.at(AxesEnum::Z)].as<double>());
+
+        // Read power multiplier
+        power_scale_factor = config["power_scale_factor"].as<double>();
+
+        // Read wrench matrix file paths
+        wrench_matrix_file_path = config["wrench_matrix_file_path"].as<std::string>();
+        wrench_matrix_pinv_file_path = config["wrench_matrix_pinv_file_path"].as<std::string>();
+
+        // Convert wrench matrix file paths to absolute paths
+        wrench_matrix_file_path = CONTROLS_PACKAGE_PATH + "/" + wrench_matrix_file_path;
+        wrench_matrix_pinv_file_path = CONTROLS_PACKAGE_PATH + "/" + wrench_matrix_pinv_file_path;
+    } catch (const std::exception &e) {
+        ROS_ERROR("Exception: %s", e.what());
+        ROS_ASSERT_MSG(false, "Could not read robot config file. Make sure it is in the correct format. '%s'",
+                       ROBOT_CONFIG_FILE_PATH.c_str());
+    }
+}
+
+void ControlsUtils::update_robot_config(std::function<void(YAML::Node &)> update_function, std::string update_name) {
+    try {
+        // Lock the mutex to prevent other threads from accessing the robot config file while it is being updated
+        std::lock_guard<std::mutex> guard(robot_config_mutex);
+
+        // Read the robot config file into a YAML node and update it with the provided function
+        YAML::Node config = YAML::LoadFile(ROBOT_CONFIG_FILE_PATH);
+        update_function(config);
+
+        // Open the robot config file for writing
+        std::ofstream fout(ROBOT_CONFIG_FILE_PATH);
+        if (!fout.is_open()) throw;
+
+        // Set exceptions to be thrown on failure
+        fout.exceptions(std::ofstream::failbit | std::ofstream::badbit);
+
+        // Write the updated YAML node to the file
+        fout << config;
+        if (fout.fail()) throw;
+
+        fout.close();
+    } catch (const std::exception &e) {
+        ROS_ERROR("Exception: %s", e.what());
+        ROS_ASSERT_MSG(false, "Could not update %s in robot config file. Make sure it is in the correct format. '%s'",
+                       update_name.c_str(), ROBOT_CONFIG_FILE_PATH.c_str());
+    }
+}
+
+void ControlsUtils::update_robot_config_pid_gains(const LoopsMap<AxesMap<PIDGainsMap>> &loops_axes_pid_gains,
```

Comment:
In each of these 3 update robot config functions, instead of defining a sub-function and calling `update_robot_config` with `update_function`, can we write separate functions to open and close the robot config file (with mutex handling) instead? The function flow is open function -> specific update logic -> close function.

Reply:
I agree that this would make the functions easier to understand. However, one of the motivations behind this design was to avoid having to copy-paste the `try`/`catch` block in `update_robot_config`, especially the errors in the `catch`. Your proposal would require copy-pasting the `try`/`catch` block into each update function. I am open to considering an alternative proposal that avoids this issue, but for now, I think the solution I implemented is best. Yes, it takes a bit to understand what's going on, but it's not a huge deal for a future developer to spend a few minutes understanding it.